homeassistant/components/godice/sensor.py

```diff
@@ -0,0 +1,136 @@
+"""Sensors for GoDice device."""
+
+import logging
+
+import godice
+
+from homeassistant import const
+from homeassistant.components.sensor import (
+    RestoreSensor,
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DATA_DEVICE, DATA_DEVICE_INFO, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+ROLLED_NUMBER_SENSOR_DESCR = SensorEntityDescription(
+    key="rolled_number", state_class=SensorStateClass.MEASUREMENT
+)
+COLOR_SENSOR_DESCR = SensorEntityDescription(
+    key="color",
+    device_class=SensorDeviceClass.ENUM,
+    options=[color.name for color in godice.Color],
+)
+BATTERY_SENSOR_DESCR = SensorEntityDescription(
+    key="battery_level",
+    device_class=SensorDeviceClass.BATTERY,
+    native_unit_of_measurement=const.PERCENTAGE,
+    state_class=SensorStateClass.MEASUREMENT,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Setups Dice sensors."""
+    data = hass.data[DOMAIN][config_entry.entry_id]
+    device = data[DATA_DEVICE]
+    device_info = data[DATA_DEVICE_INFO]
+
+    entry_ctors = [
+        DiceColorSensor,
+        BatteryLevelSensor,
+        DiceNumberSensor,
+    ]
+
+    entries = [ctor(device_info, device) for ctor in entry_ctors]
+    async_add_entities(entries)
+
+
+class BaseSensor(SensorEntity):
+    """Base for concrete sensors."""
+
+    _attr_should_poll = False
+    _attr_has_entity_name = True
+
+    def __init__(
+        self, entity_description: SensorEntityDescription, device_info: DeviceInfo
+    ) -> None:
+        """Set default values."""
+        self.entity_description = entity_description
+        self._attr_device_info = device_info
+        self._attr_unique_id = (
+            f"{device_info[const.CONF_NAME]}_{self.entity_description.key}"
+        )
+        self._attr_translation_key = entity_description.key
+
+
+class DiceColorSensor(RestoreSensor, BaseSensor):
+    """Represents color of a dice (dots)."""
+
+    dice: godice.Dice
+
+    def __init__(self, device_info: DeviceInfo, device: godice.Dice) -> None:
+        """Set default values."""
+        super().__init__(COLOR_SENSOR_DESCR, device_info)
+        self.dice = device
```

Comment:
DiceColorSensor, BatteryLevelSensor and DiceNumberSensor all have a `dice` class variable. Why don't you put it in BaseSensor? Then you can also remove the __init__ for each of these right?

And you can include the descriptions like below. Also small point but you write `async_add_entities(entries)` but you are really adding entities instead of entries, but that won't be a problem if you do it like this.
```python
    entities = [
        (DiceColorSensor, COLOR_SENSOR_DESCR),
        (BatteryLevelSensor, BATTERY_SENSOR_DESCR),
        (DiceNumberSensor, ROLLED_NUMBER_SENSOR_DESCR),
    ]
    async_add_entities([entity_class(entity_description, device_info, device) for entity_class, entity_description in entities])
```

Reply:
good catch, really better to adjust a base constructor to be used for all child classes. Done